app/services/onestop_id.rb

```diff
@@ -1,123 +1,153 @@
-      errors << "must start with \"#{ENTITY_TO_PREFIX.values.join(' or ')}\" as its 1st component"
-      is_a_valid_onestop_id = false
+    def validate_suffix(value)
+      true
     end
 
-    if onestop_id.split(COMPONENT_SEPARATOR)[1].length == 0 || !valid_component?(:geohash, onestop_id.split(COMPONENT_SEPARATOR)[1])
-      errors << 'must include a valid geohash as its 2nd component'
-      is_a_valid_onestop_id = false
+    private
+
+    def name_filter(value)
+      value.downcase.gsub(NAME_TILDE, '~').gsub(NAME_FILTER, '')
     end
 
-    if !valid_component?(:name, onestop_id.split(COMPONENT_SEPARATOR)[2])
-      errors << 'must include only letters, digits, and ~ or @ in its abbreviated name (the 3rd component)'
-      is_a_valid_onestop_id = false
+    def geohash_filter(value)
+      value.downcase.gsub(GEOHASH_FILTER, '')
     end
-    return is_a_valid_onestop_id, errors
   end
 
-  def self.find(onestop_id)
-    OnestopId::PREFIX_TO_MODEL[onestop_id.split(OnestopId::COMPONENT_SEPARATOR)[0]].find_by(onestop_id: onestop_id)
+  class OperatorOnestopId < OnestopIdBase
+    PREFIX = :o
+    MODEL = Operator
   end
 
-  def self.find!(onestop_id)
-    OnestopId::PREFIX_TO_MODEL[onestop_id.split(OnestopId::COMPONENT_SEPARATOR)[0]].find_by!(onestop_id: onestop_id)
+  class FeedOnestopId < OnestopIdBase
+    PREFIX = :f
+    MODEL = Feed
+  end
+
+  class StopOnestopId < OnestopIdBase
+    PREFIX = :s
+    MODEL = Stop
   end
 
-  private
+  class RouteOnestopId < OnestopIdBase
+    PREFIX = :r
+    MODEL = Route
+  end
 
-  def name_filter(value)
-    value.downcase.gsub(NAME_TILDE, '~').gsub(NAME_FILTER, '')
+  class RouteStopPatternOnestopId < OnestopIdBase
+    PREFIX = :r
+    # MODEL = RouteStopPattern
+    NUM_COMPONENTS = 4
   end
 
-  def geohash_filter(value)
-    value.downcase.gsub(GEOHASH_FILTER, '')
+  LOOKUP = Hash[OnestopId::OnestopIdBase.descendants.map { |c| [[c::PREFIX, c::NUM_COMPONENTS], c] }]
+
+  def self.lookup(string: nil, prefix: nil, num_components: 3)
+    if string && string.length > 0
+      split = string.split(COMPONENT_SEPARATOR)
+      prefix = split[0]
+      prefix = prefix.to_sym if prefix
+      num_components = split.size
+    end
+    prefix = prefix.to_sym if prefix
+    LOOKUP[[prefix, num_components]]
   end
 
-  def self.valid_component?(component, value)
-    return false if !value || value.length == 0
-    case component
-    when :entity_prefix
-      ENTITY_TO_PREFIX.values.include?(value)
-    when :geohash
-      !(value =~ GEOHASH_FILTER)
-    when :name
-      !(value =~ NAME_FILTER)
+  def self.create_identifier(feed_onestop_id, entity_prefix, entity_id)
+    IDENTIFIER_TEMPLATE.expand(
+      feed_onestop_id: feed_onestop_id,
+      entity_prefix: entity_prefix,
+      entity_id: entity_id
+    ).to_s
+  end
+
+  def self.validate_onestop_id_string(onestop_id)
+    klass = lookup(string: onestop_id)
+    return false, ['must not be empty'] if onestop_id.blank?
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.